src/transport_picoquic.cpp

```diff
@@ -0,0 +1,828 @@
+      log_msg.str("");
+      log_msg << "Closing connection stream " << stream_id;
+      transport->logger.log(LogLevel::info, log_msg.str());
+
+      if (stream_id == 0) { // close connection
+        picoquic_close(cnx, 0);
+
+      } else {
+        picoquic_set_callback(cnx, NULL, NULL);
+      }
+
+      transport->deleteStreamContext((uint64_t)cnx, stream_id);
+
+
+      return 0;
+    }
+
+    case picoquic_callback_application_close:
+    case picoquic_callback_close: {
+      log_msg.str("");
+      log_msg << "Closing connection stream_id: " << stream_id;
+
+      if (stream_cnx != NULL) {
+        log_msg << stream_cnx->peer_addr_text;
+      }
+
+      transport->logger.log(LogLevel::info, log_msg.str());
+
+      picoquic_set_callback(cnx, NULL, NULL);
+      picoquic_close(cnx, 0);
+      transport->deleteStreamContext((uint64_t)cnx, stream_id);
+
+      if (not transport->isServerMode) {
+        transport->setStatus(TransportStatus::Disconnected);
+        return PICOQUIC_NO_ERROR_TERMINATE_PACKET_LOOP;
+      }
+
+      return 0;
+    }
+
+    case picoquic_callback_ready: { // Connection callback, not per stream
+      if (stream_cnx == NULL) {
+        stream_cnx = transport->createStreamContext(cnx, stream_id);
+      }
+
+      if (transport->isServerMode) {
+
+        // Create new stream context for new stream/connection
+        picoquic_enable_keep_alive(cnx, 3000000);
+        (void)picoquic_mark_datagram_ready(cnx, true);
+
+        transport->on_new_connection(stream_cnx);
+      }
+
+      else {
+        // Client
+        transport->setStatus(TransportStatus::Ready);
+        log_msg << "Connection established to server " << stream_cnx->peer_addr_text
+                << " stream_id: " << stream_id;
+        transport->logger.log(LogLevel::info, log_msg.str());
+      }
+
+      break;
+    }
+
+    case picoquic_callback_prepare_to_send:
+    {
+      transport->sendOutData(stream_cnx, bytes, length);
+      break;
+    }
+
+    default:
+      log_msg.str("");
+      log_msg << "Got event " << fin_or_event;
+      transport->logger.log(LogLevel::debug, log_msg.str());
+      break;
+  }
+
+  return 0;
+}
+
+int pq_loop_cb(picoquic_quic_t* quic, picoquic_packet_loop_cb_enum cb_mode,
+               void* callback_ctx, void* callback_arg)
+{
+
+  PicoQuicTransport *transport = (PicoQuicTransport *)callback_ctx;
+  int ret = 0;
+  std::ostringstream log_msg;
+
+  if (transport == NULL) {
+    return PICOQUIC_ERROR_UNEXPECTED_ERROR;
+  }
+  else {
+    log_msg << "Loop got cb_mode: ";
+
+    switch (cb_mode) {
+      case picoquic_packet_loop_ready: {
+        log_msg << "packet_loop_ready, waiting for packets";
+        transport->logger.log(LogLevel::info, log_msg.str());
+
+        if (transport->isServerMode)
+          transport->setStatus(TransportStatus::Ready);
+
+        if (callback_arg != nullptr) {
+          auto *options = (picoquic_packet_loop_options_t *) callback_arg;
```

Comment:
Should prefer explicit casting to c-style casting.

Reply:
I can update them, but some of them require ```reinterpret_cast```.  It gets a bit more verbose, but it should be fine.